### ISSUE TITLE
Please also support AWS Launch Templates

### ISSUE THREAD

**Issue by pschulten** [Quentin-M/etcd-cloud-operator#55]
If you switch from Launch Configs  to Launch Templates the operator fails with:
```
	Time                            aws_private_ip  SYSTEMD_UNIT    MESSAGE
	May 7, 2020 @ 20:26:25.247	xx.xx.xxx.116	eco.service	github.com/quentin-m/etcd-cloud-operator/pkg/operator.fetchStatuses(0xc00081c000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0xc0001f0280)
	May 7, 2020 @ 20:26:25.247	xx.xx.xxx.116	eco.service		/go/src/github.com/quentin-m/etcd-cloud-operator/pkg/operator/misc.go:120 +0x406
	May 7, 2020 @ 20:26:25.247	xx.xx.xxx.116	eco.service		/go/src/github.com/quentin-m/etcd-cloud-operator/pkg/operator/operator.go:134 +0x313
	May 7, 2020 @ 20:26:25.247	xx.xx.xxx.116	eco.service		/go/src/github.com/quentin-m/etcd-cloud-operator/cmd/operator/operator.go:61 +0x4ba
	May 7, 2020 @ 20:26:25.247	xx.xx.xxx.116	eco.service		/go/src/github.com/quentin-m/etcd-cloud-operator/pkg/operator/operator.go:103 +0x51
	May 7, 2020 @ 20:26:25.247	xx.xx.xxx.116	eco.service	main.main()
	May 7, 2020 @ 20:26:25.247	xx.xx.xxx.116	eco.service	github.com/quentin-m/etcd-cloud-operator/pkg/operator.(*Operator).evaluate(0xc00038a180, 0x0, 0x0)
	May 7, 2020 @ 20:26:25.247	xx.xx.xxx.116	eco.service	github.com/quentin-m/etcd-cloud-operator/pkg/operator.(*Operator).Run(0xc00038a180)
	May 7, 2020 @ 20:26:25.246	xx.xx.xxx.116	eco.service	panic: runtime error: index out of range [-1]
	May 7, 2020 @ 20:26:25.246	xx.xx.xxx.116	eco.service	goroutine 1 [running]:
	May 7, 2020 @ 20:26:25.244	xx.xx.xxx.116	eco.service	time="2020-05-07T18:26:25Z" level=warning msg="failed to create etcd cluster client" error="etcdclient: no available endpoints"
	May 7, 2020 @ 20:26:25.112	xx.xx.xxx.116	eco.service	2020-05-07 18:26:25.111773 W | rafthttp: lost the TCP streaming connection with peer 508fa2d34d1b809d (stream Message writer)
	May 7, 2020 @ 20:26:24.751	xx.xx.xxx.116	eco.service	2020-05-07 18:26:24.750870 W | rafthttp: lost the TCP streaming connection with peer 508fa2d34d1b809d (stream MsgApp v2 reader)
	May 7, 2020 @ 20:26:24.751	xx.xx.xxx.116	eco.service	2020-05-07 18:26:24.751207 E | rafthttp: failed to read 508fa2d34d1b809d on stream MsgApp v2 (unexpected EOF)
	May 7, 2020 @ 20:26:24.751	xx.xx.xxx.116	eco.service	2020-05-07 18:26:24.751696 W | rafthttp: lost the TCP streaming connection with peer 508fa2d34d1b809d (stream Message reader)
	May 7, 2020 @ 20:26:24.728	xx.xx.xxx.62	eco.service	github.com/quentin-m/etcd-cloud-operator/pkg/operator.(*Operator).Run(0xc000001c80)
	May 7, 2020 @ 20:26:24.728	xx.xx.xxx.62	eco.service		/go/src/github.com/quentin-m/etcd-cloud-operator/pkg/operator/operator.go:134 +0x313
	May 7, 2020 @ 20:26:24.728	xx.xx.xxx.62	eco.service		/go/src/github.com/quentin-m/etcd-cloud-operator/pkg/operator/operator.go:103 +0x51
	May 7, 2020 @ 20:26:24.728	xx.xx.xxx.62	eco.service		/go/src/github.com/quentin-m/etcd-cloud-operator/cmd/operator/operator.go:61 +0x4ba
	May 7, 2020 @ 20:26:24.728	xx.xx.xxx.62	eco.service	main.main()
	May 7, 2020 @ 20:26:24.727	xx.xx.xxx.62	eco.service	github.com/quentin-m/etcd-cloud-operator/pkg/operator.(*Operator).evaluate(0xc000001c80, 0x0, 0x0)
	May 7, 2020 @ 20:26:24.727	xx.xx.xxx.62	eco.service	github.com/quentin-m/etcd-cloud-operator/pkg/operator.fetchStatuses(0xc0008c6900, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0xc0002e4880)
	May 7, 2020 @ 20:26:24.727	xx.xx.xxx.62	eco.service		/go/src/github.com/quentin-m/etcd-cloud-operator/pkg/operator/misc.go:120 +0x406
	May 7, 2020 @ 20:26:24.726	xx.xx.xxx.62	eco.service	panic: runtime error: index out of range [-1]
	May 7, 2020 @ 20:26:24.726	xx.xx.xxx.62	eco.service	goroutine 1 [running]:
	May 7, 2020 @ 20:26:24.724	xx.xx.xxx.62	eco.service	time="2020-05-07T18:26:24Z" level=warning msg="failed to create etcd cluster client" error="etcdclient: no available endpoints"
	May 7, 2020 @ 20:26:24.511	xx.xx.xxx.116	eco.service	2020-05-07 18:26:24.509500 W | rafthttp: lost the TCP streaming connection with peer 859d080ca38f4cc1 (stream MsgApp v2 writer)
	May 7, 2020 @ 20:26:21.861	xx.xx.xxx.116	eco.service	2020-05-07 18:26:21.861385 W | etcdserver: cannot get the version of member 859d080ca38f4cc1 (Get https://xx.xx.xxx.77:2380/version: dial tcp xx.xx.xxx.77:2380: connect: connection refused)
	May 7, 2020 @ 20:26:21.861	xx.xx.xxx.116	eco.service	2020-05-07 18:26:21.861362 W | etcdserver: failed to reach the peerURL(https://xx.xx.xxx.77:2380) of member 859d080ca38f4cc1 (Get https://xx.xx.xxx.77:2380/version: dial tcp xx.xx.xxx.77:2380: connect: connection refused)
	
```

thank you very much for this NICE project!

**Comment by Quentin-M** [comment]
Hi @pschulten,

Thanks for the report~ I hope the project can be as useful as it is for us to as many people as possible. It's been amazing to support our trading exchange.

Given you are seeing this after switching from AWS Launch Configuration to AWS Launch Templates, I would bet that the issue lies in https://github.com/Quentin-M/etcd-cloud-operator/blob/master/pkg/providers/asg/aws/aws.go.

Obviously [misc.go#L120](https://github.com/Quentin-M/etcd-cloud-operator/blob/64842826a4dcd90170c58d53d744a7754d461b1b/pkg/operator/misc.go#L120) is not designed to see a length of 0 instances (the whole operator would not run if there was no instances in the ASG!). Given you must have changed the Terraform to use AWS Launch Templates, would you mind getting on your instance(s) and trying to reproduce the logic that aws.go above executes (i.e. call the metadata API to find the ASG the instance is currently in, describing the ASG and then describing the instances/reservations within that ASG). I think that the AWS responses may differ slightly, making the function return no instances (instead of returning the instances ID + Private IPs of the ASG). Once the function returns the list of instances properly, the operator will start working again! :)

**Comment by pschulten** [comment]
Thanks for the reply!
I'm pretty new to go, but I'll try my best (please be lenient towards me ;))

**Comment by pschulten** [comment]
Sorry for the noise. The AWS ASG provider runs perfectly fine with both, Launch Templates and Launch Configs. The error was in my surrounding setup. Any plans for an official release? The ACL addition is superb for secure Prometheus scraping.

**Comment by Quentin-M** [comment]
Dear @pschulten,

I am super glad you were able to figure it out, have fun!
I just released latest as `qmachu/etcd-cloud-operator:v3.4.4a`, good idea~
